src/projection.jl

```diff
@@ -0,0 +1,62 @@
+using LinearAlgebra: Diagonal, diag
+
+"""
+    project([T::Type], x, dx)
+
+"project" `dx` onto type `T` such that it is the same size as `x`. If `T` is not provided,
+it is assumed to be the type of `x`.
+
+It's necessary to have `x` to ensure that it's possible to project e.g. `AbstractZero`s
+onto `Array`s -- this wouldn't be possible with type information alone because the neither
+`AbstractZero`s nor `T` know what size of `Array` to produce.
+"""
+function project end
+
+project(x, dx) = project(typeof(x), x, dx)
+
+# identity
+project(::Type{T}, x::T, dx::T) where T = dx
+
+### AbstractZero
+project(::Type{T}, x::T, dx::AbstractZero) where T = zero(x)
+
+### AbstractThunk
+project(::Type{T}, x::T, dx::AbstractThunk) where T = project(x, unthunk(dx))
+
+
+### Number-types
+project(::Type{T}, x::T, dx::T2) where {T<:Number, T2<:Number} = T(dx)
+project(::Type{T}, x::T, dx::Complex) where {T<:Real} = T(real(dx))
+
+
+# Arrays
```

Comment:
I have a feeling we might need to support the same kwargs that `similar` does.
which iirc are `size` and `eltype`.
So you can say this AbstractArray type, except the the size should be this and the eltype that.

E.g. 
for `primal` being `WrapperArray{Tuple{Float64,Float64}}`
```
project(typeof(primal), dx; eltype=Tangent{Tuple{Float64,Float64}}
```
to get back a `WrapperArray{Tangent{Tuple{Float64,Float64}}}`

I am not actualyl sure we need `size`.

We might need a `structual_zeros_from`  for if structural zero locations are only known by value (like for `SparseCSC`)

Reply:
I can see where you're coming from, but is it not both cleaner _and_ more general to do the kind of thing that @CarloLucibello suggests?

Reply:
these seem orthogonal to me.
Or at least that the closure @CarloLucibello suggest would be implemented on top of this.
So that it extracts these bits of information from the primal and passes them into the kwargs.
Rather than closing over everything

Reply:
> these seem orthogonal to me.

Hmmm maybe I've misunderstood your reasoning for needing `eltype` and `size` kwargs then.

Are you saying we might need it because we would want to apply `project` for some primal other than the primal that the tangent is for?

i.e. let `x` and `y` be primals of differing `size`s. Say that I've got a tangent `dx` for `x`, but for some reason I've only got access to `y`. So I say `project(y, dx; size=size_of_x)`.

Is this the kind of situation you're imagining?